pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,119 @@
+package writelimiter
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+// ErrResourcesExhausted is returned when no more BlobWriters can be created
+// because of the limit.
+var ErrResourcesExhausted = errors.New("writers limit is reached")
+
+type writerBouncer struct {
+	sem  *semaphore
+	once sync.Once
+}
+
+func newWriterBouncer(sem *semaphore) (*writerBouncer, error) {
+	if !sem.TryDown() {
+		return nil, ErrResourcesExhausted
```

Comment:
It is the right error. Docker stops retrying on [429 errcode.ErrorCodeTooManyRequests](https://github.com/moby/moby/blob/54251b53d7081f1471c0a2cb23fd4e198e71bd14/distribution/errors.go#L135), from the Docker point of view it doesn't means "try again", but "please stop, I can't handle this".